spark_log_parser/parsing_models/application_model_v2.py

```diff
@@ -1,131 +1,505 @@
-            val1 = ConfigValidationEMR(app=appobj, debug=debug)
-        elif appobj.cloud_platform == "databricks":
-            val1 = ConfigValidationDatabricks(app=appobj, debug=debug)
+        # Extract bucket and key from s3 filepath
+        path = filepath.replace("s3://", "").split("/")
+        bucket = path[0]
+        key = ("/".join(path[1:])).lstrip("/") + ".json"
 
-        val1.validate()
+        if compress is False:
+            s3.put_object(Bucket=bucket, Body=json.dumps(saveDat).encode("utf-8"), Key=key)
+        else:
+            dat = gzip.compress(json.dumps(saveDat).encode("utf-8"))
+            s3.put_object(Bucket=bucket, Body=dat, Key=key + ".gz")
 
-        val2 = EventDataValidation(app=appobj, debug=debug)
-        val2.validate()
+        logging.info("Saved object to cloud: %s" % (key))
 
-    def getRecentEvents(self):
 
-        tcomp = np.concatenate(([0.0], self.taskData["end_time"].values))
+class AbstractLocalFileDataLoader(abc.ABC, DataLoader):
+    @abc.abstractmethod
+    def read_uncompressed_file(self, file):
+        return file.read()
 
-        if self.existsSQL:
-            tcomp = np.concatenate(
-                (tcomp, self.sqlData["start_time"].values, self.sqlData["end_time"].values)
+    @abc.abstractmethod
+    def read_compressed_file(self, file):
+        return file.read().decode("ascii")
+
+    def load_from_local(self, filepath):
+        if ".gz" in filepath:
+            with gzip.open(filepath, "r") as fin:
+                save_data = self.read_compressed_file(fin)
+        else:
+            with open(filepath, "r") as fin:
+                save_data = self.read_uncompressed_file(fin)
+
+        return save_data
+
+    async def batch_load_fn(self, keys):
+        return [self.load_from_local(filepath) for filepath in keys]
+
+
+class LocalFileBlobDataLoader(AbstractLocalFileDataLoader):
+    def read_uncompressed_file(self, file):
+        return super().read_uncompressed_file(file)
+
+    def read_compressed_file(self, file):
+        return super().read_compressed_file(file)
+
+
+class LocalFileLinesDataLoader(AbstractLocalFileDataLoader):
+    def read_uncompressed_file(self, file):
+        lines = []
+        for line in file:
+            lines.append(line)
+
+        return lines
+
+    def read_compressed_file(self, file):
+        return self.read_uncompressed_file(file)
+
+
+class AbstractS3FileDataLoader(abc.ABC, DataLoader):
+    _s3 = None
+
+    @property
+    def s3(self):
+        if self._s3 is None:
+            self._s3 = boto3.resource("s3")
+
+        return self._s3
+
+    @abc.abstractmethod
+    def read_uncompressed_data(self, data):
+        return data.decode("ascii")
+
+    @abc.abstractmethod
+    def read_compressed_data(self, data):
+        uncompressed = gzip.decompress(data)
+        return uncompressed.decode("utf-8")
+
+    def load_from_s3(self, filepath):
+        # TODO - there must be some better way of parsing this?
```

Comment:
Ah, this is a nice class to know about. I've also since learned about the `ParseResult` class in `urllib`, which we actually use elsewhere in the log parser. Parsing an S3 URI with it gives us an object that is a little cleaner to use than a raw string - 

```python
>>> from urllib.parse import ParseResult, urlparse
>>> parsed = urlparse("s3://some/bucket/file/path.gz")
>>> parsed
ParseResult(scheme='s3', netloc='some', path='/bucket/file/path.gz', params='', query='', fragment='')
```

where the `netloc` field corresponds to the `bucket`, and the `path` corresponds to the `key`